intelmq/bots/parsers/phishtank/parser.py

```diff
@@ -1,58 +1,62 @@
 #
 # SPDX-License-Identifier: AGPL-3.0-or-later
 
-# -*- coding: utf-8 -*-
-import csv
-import io
 
-from intelmq.lib import utils
+import dateutil.parser
+
 from intelmq.lib.bot import ParserBot
 
 
 class PhishTankParserBot(ParserBot):
-    """Parse the PhishTank feed"""
-
-    def process(self):
-        report = self.receive_message()
-
-        columns = ["__IGNORE__",
-                   "source.url",
-                   "event_description.url",
-                   "time.source",
-                   "__IGNORE__",
-                   "__IGNORE__",
-                   "__IGNORE__",
-                   "event_description.target"
-                   ]
-
-        raw_report = utils.base64_decode(report.get("raw"))
-        raw_report = raw_report.translate({0: None})
-        for row in csv.reader(io.StringIO(raw_report)):
-
-            if not len(row):  # csv module can give empty lists
-                self.acknowledge_message()
-                return
-
-            # ignore headers
-            if "phish_id" in row:
-                continue
-
-            event = self.new_event(report)
-            event.change("feed.url", event["feed.url"][:event["feed.url"].find('data/')])
-
-            for key, value in zip(columns, row):
-
-                if key == "__IGNORE__":
-                    continue
-
-                event.add(key, value)
-
-            event.add('classification.type', 'phishing')
-            event.add("raw", ",".join(row))
-
-            self.send_message(event)
-        self.acknowledge_message()
+    """
+    Parse the PhishTank feed (json)
+    List of source fields:
+    [
+        'phish_id',
+        'url',
+        'phish_detail_url',
+        'submission_time',
+        'verified',
+        'verification_time',
+        'online',
+        'target',
+        'details'
+    ]
+    """
+
+    parse = ParserBot.parse_json
+    recover_line = ParserBot.recover_line_json
+
+    def parse_line(self, line, report):
+
+        event = self.new_event(report)
+
+        if line.get("submission_time"):
+            try:
+                event.add("time.source", str(dateutil.parser.parse(line.get("submission_time"))), raise_failure=False)
+            except dateutil.parser.ParserError:
+                pass
+
+        event.add("classification.type", "phishing")
+        event.add("extra.phish_id", line.get("phish_id"), raise_failure=False)
+        event.add("source.url", line.get("url"))
+        event.add("extra.phish_detail_url", line.get("phish_detail_url"), raise_failure=False)
+        event.add("extra.verified", line.get("verified"), raise_failure=False)
+        event.add("extra.verification_time", line.get("verification_time"), raise_failure=False)
+        event.add("extra.online", line.get("online"), raise_failure=False)
+        event.add("extra.target", line.get("target"), raise_failure=False)
+
+        if line.get("details") and isinstance(line.get("details"), list) and len(line.get("details")) > 0:
```

Comment:
The problem is that this field is completely undocumented. From my observations it contains DNS resolutions of the phishing domain at various times.

All cases are:

* `details` array is empty or non-existent: This might mean that Phishtank couldn't resolve the domain for unknown reason. No additional information is added. Checking `line.get("details")` returns `None` when `details` doesn't exist and the other if-conditions are not evaluated any more.

* `details` array contains one element: The first resolution of the phishing occurs few minutes after the submission of the phishing url. It provides additional information about IP address.

* `details` array contains more than one elemenet: The first element (at index 0) is the earliest time of resolution (few minutes after url submission). Subsequent elements contains later resolutions. We are mostly interested in the resolution at the time of submission. Therefore we use the first element.